mm/kfence/kfence-test.c

```diff
@@ -0,0 +1,485 @@
+		break;
+	case KFENCE_ERROR_CORRUPTION:
+		cur += scnprintf(cur, end - cur, "Detected corrupted memory");
+		break;
+	}
+
+	cur += scnprintf(cur, end - cur, " at 0x%px", (void *)r->addr);
+
+	spin_lock_irqsave(&observed.lock, flags);
+	if (!report_available())
+		goto out; /* A new report is being captured. */
+
+	/* Finally match expected output to what we actually observed. */
+	ret = strstr(observed.lines[0], expect[0]) && strstr(observed.lines[1], expect[1]);
+out:
+	spin_unlock_irqrestore(&observed.lock, flags);
+	return ret;
+}
+
+/* ===== Test cases ===== */
+
+#define TEST_PRIV_WANT_MEMCACHE ((void *)1)
+
+/* Cache used by tests; if NULL, allocate from kmalloc instead. */
+static struct kmem_cache *test_cache;
+
+static void setup_test_cache(struct kunit *test, size_t size, void (*ctor)(void *))
+{
+	if (test->priv != TEST_PRIV_WANT_MEMCACHE) {
+		KUNIT_ASSERT_FALSE_MSG(test, ctor, "unexpected ctor?");
+		return;
+	}
+
+	/*
+	 * Use SLAB_NOLEAKTRACE to prevent merging with existing caches. Any
+	 * other flag in SLAB_NEVER_MERGE (except SLAB_TYPESAFE_BY_RCU, which
+	 * disables KFENCE) also works. Use SLAB_ACCOUNT to allocate via memcg,
+	 * if enabled.
+	 */
+	test_cache = kmem_cache_create("test", size, 1, SLAB_NOLEAKTRACE | SLAB_ACCOUNT, ctor);
+	KUNIT_ASSERT_TRUE_MSG(test, test_cache, "could not create cache");
+}
+
+/* Must always inline to match stack trace against caller. */
+static __always_inline void test_free(void *ptr)
+{
+	if (test_cache)
+		kmem_cache_free(test_cache, ptr);
+	else
+		kfree(ptr);
+}
+
+/* On which side the allocation and the closest guard page should be. */
+enum allocation_side {
+	ALLOCATE_ANY,
+	ALLOCATE_LEFT,
+	ALLOCATE_RIGHT,
+};
+
+/*
+ * Try to get a guarded allocation from KFENCE. Uses either kmalloc() or the
+ * current test_cache if set up.
+ */
+static void *test_alloc(struct kunit *test, size_t size, gfp_t gfp, enum allocation_side side)
+{
+	void *alloc;
+	unsigned long timeout;
+
+	kunit_info(test, "%s: size=%zu, gfp=%x, side=%s, cache=%i", __func__, size, gfp,
+		   side == ALLOCATE_ANY ? "any" : (side == ALLOCATE_LEFT ? "left" : "right"),
+		   !!test_cache);
+
+	/*
+	 * 10x the sample rate should be more than enough to ensure we get a
+	 * KFENCE allocation.
+	 */
+	timeout = jiffies + msecs_to_jiffies(10 * CONFIG_KFENCE_SAMPLE_RATE);
+	do {
+		if (test_cache)
+			alloc = kmem_cache_alloc(test_cache, gfp);
+		else
+			alloc = kmalloc(size, gfp);
+
+		if (is_kfence_addr(alloc)) {
+			if (side == ALLOCATE_ANY)
+				return alloc;
+			if (side == ALLOCATE_LEFT && IS_ALIGNED((unsigned long)alloc, PAGE_SIZE))
+				return alloc;
+			if (side == ALLOCATE_RIGHT && !IS_ALIGNED((unsigned long)alloc, PAGE_SIZE))
+				return alloc;
+		}
+
+		test_free(alloc);
+	} while (jiffies < timeout);
+
+	KUNIT_ASSERT_TRUE_MSG(test, false, "failed to allocate from KFENCE");
+	return NULL; /* Unreachable. */
+}
+
+static void test_out_of_bounds_read(struct kunit *test)
```

Comment:
I think we need `noinline` for the test functions.

Reply:
We pass them to KUnit as function pointers, so it will never inline any of this since the place that calls these is somewhere deep in KUnit.